From e7fdc0560ac5f1ec8c8e9c71abaed9ffd8f398f7 Mon Sep 17 00:00:00 2001
From: Rui Wang <ruwang@redhat.com>
Date: Mon, 21 Jul 2014 03:22:47 -0400
Subject: [x86] efi: Fix earlyprintk off-by-one bug

Message-id: <1405912969-15631-3-git-send-email-ruwang@redhat.com>
Patchwork-id: 86046
O-Subject: [PATCH RHEL7.1 BZ1085558 2/4] backport x86/efi earlyprintk off-by-one bugfix.
Bugzilla: 1085558
RH-Acked-by: Dave Young <dyoung@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

From: "Wang, Xiaolong" <xiaolong.wang@intel.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1085558

Backport of:
Original commit 1f3a8bae21a22808f242ccfb22aca37a635e261e
Original Author: Matt Fleming <matt.fleming@intel.com>
Original Date:   Mon Nov 11 16:18:59 2013 +0000

    x86/efi: Fix earlyprintk off-by-one bug

    Dave reported seeing the following incorrect output on his Thinkpad T420
    when using earlyprintk=efi,

    [    0.000000] efi: EFI v2.00 by Lenovo
                        ACPI=0xdabfe000  ACPI 2.0=0xdabfe014 SMBIOS=0xdaa9e000

    The output should be on one line, not split over two. The cause is an
    off-by-one error when checking that the efi_y coordinate hasn't been
    incremented out of bounds.

Signed-off-by: Wang, Xiaolong <xiaolong.wang@intel.com>
Signed-off-by: Wang, Rui <ruwang@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/platform/efi/early_printk.c b/arch/x86/platform/efi/early_printk.c
index 6599a00..81b506d 100644
--- a/arch/x86/platform/efi/early_printk.c
+++ b/arch/x86/platform/efi/early_printk.c
@@ -142,7 +142,7 @@ early_efi_write(struct console *con, const char *str, unsigned int num)
    efi_y += font->height;
   }
 
-  if (efi_y + font->height >= si->lfb_height) {
+  if (efi_y + font->height > si->lfb_height) {
    u32 i;
 
    efi_y -= font->height;
-- 
1.7.1